core/state/snapshot/iterator_test.go

```diff
@@ -32,19 +32,15 @@ import (
 // TestAccountIteratorBasics tests some simple single-layer(diff and disk) iteration
 func TestAccountIteratorBasics(t *testing.T) {
 	var (
-		destructs = make(map[common.Hash]struct{})
-		accounts  = make(map[common.Hash][]byte)
-		storage   = make(map[common.Hash]map[common.Hash][]byte)
+		accounts = make(map[common.Hash][]byte)
+		storage  = make(map[common.Hash]map[common.Hash][]byte)
 	)
 	// Fill up a parent
 	for i := 0; i < 100; i++ {
 		h := randomHash()
 		data := randomAccount()
 
 		accounts[h] = data
-		if rand.Intn(4) == 0 {
```

Comment:
Somethings a bit weird here, if I re-add this and we mark every ~4th element as destructed by setting it nil, only the `difflayer.AccountIterator` will spit out 100 elements, while the other two iterators will spit out ~75.

e.g. this tests passes some of the time:

```diff
diff --git a/core/state/snapshot/iterator_test.go b/core/state/snapshot/iterator_test.go
index b9fe370b86..dff8c86ab6 100644
--- a/core/state/snapshot/iterator_test.go
+++ b/core/state/snapshot/iterator_test.go
@@ -41,6 +41,9 @@ func TestAccountIteratorBasics(t *testing.T) {
                data := randomAccount()
 
                accounts[h] = data
+               if rand.Intn(4) == 0 {
+                       accounts[h] = nil
+               }
                if rand.Intn(2) == 0 {
                        accStorage := make(map[common.Hash][]byte)
                        value := make([]byte, 32)
@@ -55,11 +58,11 @@ func TestAccountIteratorBasics(t *testing.T) {
        verifyIterator(t, 100, it, verifyNothing) // Nil is allowed for single layer iterator
 
        it = diffLayer.newBinaryAccountIterator(common.Hash{})
-       verifyIterator(t, 100, it, verifyNothing) // Nil is allowed for single layer iterator
+       verifyIterator(t, 80, it, verifyNothing) // Nil is allowed for single layer iterator
 
        diskLayer := diffToDisk(diffLayer)
        it = diskLayer.AccountIterator(common.Hash{})
-       verifyIterator(t, 100, it, verifyNothing) // Nil is allowed for single layer iterator
+       verifyIterator(t, 80, it, verifyNothing) // Nil is allowed for single layer iterator
 }
 ```

Reply:
An AccountIterator will yield every element in the layer. 
A fast- or binary- iterator will only yield non-nil elements. Importantly: any `nil`-accounts in the upper layers will _mask_ the corresponding account in the lower layers. 

Some background: a fast- or binary- will be used to serve snap requests, where we want to ship a slice of accounts, and deleted ones. If an account has been recently deleted, we want to just ignore that one.

Reply:
Further background: 

A binary iterator is the simplest "layered" iterator. It has two notion: "`this` layer and the `parent` layer". It will always "prefer" `this` layer. In order to model `N` layers, we simply make `N` binary iterators, where each parent is another binary iterator. This is very slow, naturally, and used only as a simple-but-correct reference implementation. 

The fast iterator is much more advanced, and a lot faster.